snuba/utils/streams/consumer.py

```diff
@@ -516,6 +490,23 @@ def build_kafka_consumer_configuration(
     }
 
 
+@dataclass(frozen=True)
+class Commit:
```

Comment:
I have a few questions about the goal of this class beyond what is in the PR description:
- Do you envision all Kafka payloads to be abstracted through a data class that has an encode method to generate the payload and a decode method to rebuild the class from the payload ?
- If yes, do we actually need Payload ? Or do you expect people to decide when to use directly Payload and when to build a dataclass on top of it ?
- Who is supposed to resolve the Payload into Commit when polling ? Is the consumer somehow going to automate part of that or is it part of the responsibilities of who uses the consumer ?

Reply:
> * Do you envision all Kafka payloads to be abstracted through a data class that has an encode method to generate the payload and a decode method to rebuild the class from the payload ?
> * If yes, do we actually need Payload ? Or do you expect people to decide when to use directly Payload and when to build a dataclass on top of it ?
> * Who is supposed to resolve the Payload into Commit when polling ? Is the consumer somehow going to automate part of that or is it part of the responsibilities of who uses the consumer ?

I think I can answer all three of these questions at once. Some of this is reiterating the above (hopefully a bit clearer) and some of it goes into more detail specific to this question.

My thinking (and based on this, it might be easier just do to it as part of this pull request rather than as a follow up) is that both `Consumer` and `Message` would be made generic on `TPayload`, e.g.

```python
class Consumer(Generic[TPayload], ABC):
    def poll(self, timeout: Optional[float] = None) -> Optional[Message[TPayload]]:
        …
```

I also plan on adding a `Codec` type, with this signature:

```python
class Codec(Generic[TEncoded, TDecoded], ABC):
    def encode(self, value: TDecoded) -> TEncoded:
        …
    
    def decode(self, value: TEncoded) -> TDecoded:
        …
```

The `KafkaConsumer` will have a required constructor argument, like this:

```python
class KafkaConsumer(Consumer[TPayload]):
    def __init__(
        self,
        configuration: Mapping[str, Any],
        codec: Codec[Payload, TPayload],       
        *,
        commit_retry_policy: Optional[RetryPolicy] = None,
    ) -> None:
```

`poll` will change slightly:

```diff
        result = Message(
            Partition(Topic(message.topic()), message.partition()),
            message.offset(),
-           Payload(message.key(), message.value()),
+           self.__codec.decode(Payload(message.key(), message.value())),
            datetime.fromtimestamp(message.timestamp()[1] / 1000.0),  # XXX: tzinfo?
        )
```

I think that in practice, most implementation should not use `Payload` directly (e.g. via `Codec[Payload, Payload]`, though I suppose nothing is preventing them from doing that — it might be a useful migration strategy, even.

Later, I expect that we'll add a `Producer[TPayload]`, kind of like the following or similar (it's probably missing arguments, but you get the idea):

```python
class Producer(Generic[TPayload], ABC):
    def produce(self, destination: Union[Topic, Partition], payload: TPayload) -> None:
        …
```

Reply:
ok, thanks. This makes sense

Reply:
Did all of these changes (except `Producer`) here: https://github.com/getsentry/snuba/compare/5f34e0891720ac64c066e8613bf07aaa851abe4b...c785653ce0ff801155e5749414d5b6a8ab54fdc9